mpl_interactions/controller.py

```diff
@@ -85,91 +102,81 @@ def add_kwargs(self, kwargs, slider_formats=None, play_buttons=None, allow_dupli
-                        play_button=_play_buttons[k],
-                    )
-                    if control:
-                        self.controls[k] = control
-                        self.vbox.children = list(self.vbox.children) + [control]
-                if k == "vmin_vmax":
-                    self.params["vmin"] = self.params["vmin_vmax"][0]
-                    self.params["vmax"] = self.params["vmin_vmax"][1]
+
+        if not self.use_ipywidgets:
+            axes, fig = maybe_create_mpl_controls_axes(kwargs)
+            if fig is not None:
+                self.control_figures.append((fig))
         else:
-            if len(kwargs) > 0:
-                mpl_layout = create_mpl_controls_fig(kwargs)
-                self.control_figures.append(mpl_layout[0])
-                widget_y = 0.05
-                for k, v in kwargs.items():
-                    if k in self.params:
-                        if allow_duplicates:
-                            continue
-                        else:
-                            raise ValueError("Can't overwrite an existing param in the controller")
-                    self.params[k], control, cb, widget_y = kwarg_to_mpl_widget(
-                        mpl_layout[0],
-                        mpl_layout[1:],
-                        widget_y,
-                        k,
-                        v,
-                        partial(self.slider_updated, key=k),
-                        self.slider_format_strings[k],
-                    )
-                    if control:
-                        self.controls[k] = control
-                    if k == "vmin_vmax":
-                        self.params["vmin"] = self.params["vmin_vmax"][0]
-                        self.params["vmax"] = self.params["vmin_vmax"][1]
-
-    def _slider_updated(self, change, key, values):
+            axes = [None] * len(kwargs)
+
+        for k, v in kwargs.items():
+            if k in self.params:
+                if allow_duplicates:
+                    continue
+                else:
+                    raise ValueError("can't overwrite an existing param in the controller")
+            # TODO: accept existing mpl widget
+            # if isinstance(v, AxesWidget):
+            #     self.params[k], self.controls[k], _ = process_mpl_widget(
+            #         v, partial(self.slider_updated, key=k)
+            #     )
+            # else:
+            ax = axes.pop()
+            control = kwarg_to_widget(k, v, ax, play_button=_play_buttons[k])
+            # TODO: make the try except silliness less ugly
+            # the complexity of hiding away the val vs value vs whatever needs to
+            # be hidden away somewhere - but probably not here
+            if k in index_kwargs:
+                self.params[k] = control.index
+                try:
+                    control.observe(partial(self._slider_updated, key=k), names="index")
+                except AttributeError:
```

Comment:
make this a more nuanced attribute error wher eyou look for a specific message - if thts possible